File: 19.删除链表的倒数第-n-个结点.py
#
# @lc app=leetcode.cn id=19 lang=python3
#
# [19] 删除链表的倒数第 N 个结点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # 第一种，放入栈中，pop去拿到n的前一位,但是这个空间复杂度为o~N
    # def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        # stack = list()
        # first = ListNode(0, head)
        # cur = first
        # while cur:
        #     stack.append(cur)
        #     cur = cur.next
        
        # for i in range(n):
        #     stack.pop()
        
        # prev = stack[-1]
        # prev.next = prev.next.next
        # return first.next

    # 第二种，快慢指针。先走n步，第二个指针开始走，一直到第一个指针走到最后一个。保持这两个指针为n
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        first = ListNode(0, head)
        low, high = first, first
        # 设置中间的间隔+1，因为要找到的是前一个节点
        for _ in range(n):
            high = high.next

        # while到最后一个节点
        while high.next:
            low = low.next
            high = high.next

        low.next = low.next.next
        return first.next



# @lc code=end

File: 19.删除链表的倒数第-n-个结点.go
/*
 * @lc app=leetcode.cn id=19 lang=golang
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	var first *ListNode = &ListNode{0, head}
	low, high := first, first

	// 先走块的节点,设置好间隔,注意，要找的是前n+1个节点
	for i := 0; i < n; i++ {
		high = high.Next
	}

	for ; high.Next != nil; high = high.Next {
		low = low.Next
	}
	low.Next = low.Next.Next
	return first.Next
}

// @lc code=end

